src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs

```diff
@@ -3381,10 +3381,23 @@ private BoundExpression BindNullCoalescingOperator(BinaryExpressionSyntax node,
                 return GenerateNullCoalescingBadBinaryOpsError(node, leftOperand, rightOperand, Conversion.NoConversion, diagnostics);
             }
 
-            // SPEC: Otherwise, if A exists and is a non-nullable value type, a compile-time error occurs.
-            if ((object)optLeftType != null && optLeftType.IsValueType && !optLeftType.IsNullableType())
+            // SPEC: Otherwise, if A exists and is a non-nullable value type, a compile-time error occurs. First we check for the pre-C# 8.0
+            // SPEC: condition, to ensure that we don't allow previously illegal code in old language versions.
+            if ((object)optLeftType != null && !optLeftType.IsReferenceType && !isLeftNullable)
             {
-                return GenerateNullCoalescingBadBinaryOpsError(node, leftOperand, rightOperand, Conversion.NoConversion, diagnostics);
+                // Prior to C# 8.0, the spec said that the left type must be either a reference type or a nullable value type. This was relaxed
+                // with C# 8.0, so if the feature is not enabled then issue a diagnostic and return
+                if (!optLeftType.IsValueType && !isLeftNullable)
```

Comment:
> && !isLeftNullable [](start = 44, length = 19)

Is this condition redundant?

Reply:
> Same answer as in VB. This is not redundant as unconstrained type parameters will hit this.

The question is about ``` && !isLeftNullable``` part

---
In reply to: [217216067](https://github.com/dotnet/roslyn/pull/29724#discussion_r217216067) [](ancestors = 217216067,217212991)